File: ConjugatePriorTools/dirichletMultinomialEstimation.py
#!/usr/bin/python
#
# A library for finding the optimal dirichlet prior from counts
# By: Max Sklar
# @maxsklar
# https://github.com/maxsklar

# Copyright 2013 Max Sklar

import math
import random
import logging
import scipy.special as mathExtra

def digamma(x): return float(mathExtra.psi(x))
def trigamma(x): return float(mathExtra.polygamma(1, x))

#Find the log probability that we see a certain set of data
# give our prior.mate d
def dirichLogProb(priorList, data, Beta = None, W = None):
  K = data.K
  total = 0.0
  for k in range(0, K):
    for i in range(0, len(data.U[k])):
      total += data.U[k][i]*math.log(priorList[k] + i)

  sumPrior = sum(priorList)
  for i in range(0, len(data.V)):
    total -= data.V[i] * math.log(sumPrior + i)

  # Add prior
  if (Beta != None):
    for i in range(0, K):
      total -= priorList[i]*Beta[i]

  if (W != None):
    total += W*math.lgamma(sumPrior)
    for k in range(0, K): total -= W*(math.lgamma(priorList[k]))
  return total

#Gives the derivative with respect to the prior.  This will be used to adjust the loss
def priorGradient(priorList, data, Beta = None, W = None):
	K = data.K
	
	termToSubtract = 0
	for i in range(0, len(data.V)):
		termToSubtract += float(data.V[i]) / (sum(priorList) + i)
	
	retVal = [0]*K
	for j in range(0, K):
		for i in range(0, len(data.U[j])):
			retVal[j] += float(data.U[j][i]) / (priorList[j] + i)
	
	for j in range(0, K):
		retVal[j] -= termToSubtract
	
	# Add Prior
	if (Beta != None):
		for k in range(0, K):
			retVal[k] -= Beta[k]
	
	if (W != None):
		for k in range(0, K):
			retVal[k] += W*(digamma(sum(priorList)) - digamma(priorList[k]))
		
	return retVal

#The hessian is actually the sum of two matrices: a diagonal matrix and a constant-value matrix.
#We'll write two functions to get both
def priorHessianConst(priorList, data, W = None):
	total = 0
	for i in range(0, len(data.V)):
		total += float(data.V[i]) / (sum(priorList) + i)**2
	if (W != None):
		total += W*trigamma(sum(priorList))
	return total

def priorHessianDiag(priorList, data, W = None):
  K = len(data.U)
  retVal = [0]*K
  for k in range(0, K):
    for i in range(0, len(data.U[k])):
      retVal[k] -= data.U[k][i] / (priorList[k] + i)**2
  if (W != None):
    for k in range(0, K):
      retVal[k] -= W*trigamma(priorList[k])
  return retVal

	
# Compute the next value to try here
# http://research.microsoft.com/en-us/um/people/minka/papers/dirichlet/minka-dirichlet.pdf (eq 18)
# https://www.researchgate.net/profile/Max-Sklar/publication/370927162_Algorithms_for_Multivariate_Newton-Raphson_for_Optimization/links/64698c0170202663165fd82a/Algorithms-for-Multivariate-Newton-Raphson-for-Optimization.pdf
def getPredictedStep(hConst, hDiag, gradient):
  K = len(gradient)
  numSum = sum(gradient[k] / hDiag[k] for k in range(K))
  denSum = sum(1.0 / hDiag[k] for k in range(K))
  b = numSum / ((1.0/hConst) + denSum)
  return [(b - gradient[k]) / hDiag[k] for k in range(K)]

# Uses the diagonal hessian on the log-alpha values
# https://www.researchgate.net/profile/Max-Sklar/publication/370927162_Algorithms_for_Multivariate_Newton-Raphson_for_Optimization/links/64698c0170202663165fd82a/Algorithms-for-Multivariate-Newton-Raphson-for-Optimization.pdf
def getPredictedStepAlt(hConst, hDiag, gradient, alphas):
  x = [(grad + alpha * h) for grad, alpha, h in zip(gradient, alphas, hDiag)]
  Z = 1.0 / hConst + sum(alpha / xi for alpha, xi in zip(alphas, x))
  S = sum(alpha * grad / xi for alpha, grad, xi in zip(alphas, gradient, x))
  return [(S / Z - grad) / xi for grad, xi in zip(gradient, x)]

#The priors and data are global, so we don't need to pass them in
def getTotalLoss(trialPriors, data, Beta = None, W = None):
  return -1*dirichLogProb(trialPriors, data, Beta = None, W = None)
	
def predictStepUsingHessian(gradient, priors, data, W = None):
	totalHConst = priorHessianConst(priors, data, W)
	totalHDiag = priorHessianDiag(priors, data, W)
	return getPredictedStep(totalHConst, totalHDiag, gradient)
	
def predictStepLogSpace(gradient, priors, data, W = None):
	totalHConst = priorHessianConst(priors, data, W)
	totalHDiag = priorHessianDiag(priors, data, W)
	return getPredictedStepAlt(totalHConst, totalHDiag, gradient, priors)
	

# Returns whether it's a good step, and the loss	
def testTrialPriors(trialPriors, data, Beta = None, W = None):
	for alpha in trialPriors: 
		if alpha <= 0: 
			return float("inf")
		
	return getTotalLoss(trialPriors, data, Beta, W)
	
def sqVectorSize(v):
	s = 0
	for i in range(0, len(v)): s += v[i] ** 2
	return s

class CompressedRowData:
  def __init__(self, K):
    self.K = K
    self.V = []
    self.U = []
    for k in range(0, K): self.U.append([])
    
  def appendRow(self, row, weight):
    if (len(row) != self.K): logging.error("row must have K=" + str(self.K) + " counts")
    
    for k in range(0, self.K): 
      for j in range(0, row[k]):
        if (len(self.U[k]) == j): self.U[k].append(0)
        self.U[k][j] += weight
      
    for j in range(0, sum(row)):
      if (len(self.V) == j): self.V.append(0)
      self.V[j] += weight
  

def findDirichletPriors(data, initAlphas, iterations, Beta = None, W = None):
  priors = initAlphas

  # Let the learning begin!!
  #Only step in a positive direction, get the current best loss.
  currentLoss = getTotalLoss(priors, data, Beta, W)

  gradientToleranceSq = 2 ** -10
  learnRateTolerance = 2 ** -20

  count = 0
  while(count < iterations):
    count += 1
    
    #Get the data for taking steps
    gradient = priorGradient(priors, data, Beta, W)
    gradientSize = sqVectorSize(gradient) 
    logging.debug("Iteration: %s Loss: %s ,Priors: %s, Gradient Size: %s" % (count, currentLoss, priors, gradientSize))
    
    if (gradientSize < gradientToleranceSq):
      logging.debug("Converged with small gradient")
      return priors
    
    trialStep = predictStepUsingHessian(gradient, priors, data, W)
    
    #First, try the second order method
    trialPriors = [0]*len(priors)
    for i in range(0, len(priors)): trialPriors[i] = priors[i] + trialStep[i]
    
    #TODO: Check for taking such a small step that the loss change doesn't register (essentially converged)
    #  Fix by ending
    
    loss = testTrialPriors(trialPriors, data, Beta, W)
    if loss < currentLoss:
      currentLoss = loss
      priors = trialPriors
      continue
    
    trialStep = predictStepLogSpace(gradient, priors, data, W)
    trialPriors = [0]*len(priors)
    for i in range(0, len(priors)): 
      try:
        trialPriors[i] = priors[i] * math.exp(trialStep[i])
      except:
        trialPriors[i] = priors[i]
    loss = testTrialPriors(trialPriors, data, Beta, W)

    #Step in the direction of the gradient until there is a loss improvement
    learnRate = 1.0
    while loss > currentLoss:
      learnRate *= 0.9
      trialPriors = [0]*len(priors)
      for i in range(0, len(priors)): trialPriors[i] = priors[i] + gradient[i]*learnRate
      loss = testTrialPriors(trialPriors, data, Beta, W)

    if (learnRate < learnRateTolerance):
      logging.debug("Converged with small learn rate")
      return priors

    currentLoss = loss
    priors = trialPriors
    
  logging.debug("Reached max iterations")
  return priors

File: TypescriptPort/dirichletMultinomialEstimation.ts
// Find an optimal Beta/Dirichlet distribution from count data
// Library copied from open source github.com/maxsklar/bayespy

export class CompressedRowData {
  K: number;
  V: number[];
  U: number[][];

  constructor(K: number) {
    this.K = K;
    this.V = [];
    this.U = [];

    for (let k = 0; k < K; k++) {
      this.U.push([]);
    }
  }

  appendRow(row: number[]) {
    if (row.length !== this.K) {
      console.error("row must have K=" + this.K + " counts");
      return;
    }

    for (let k = 0; k < this.K; k++) {
      for (let j = 0; j < row[k]; j++) {
        if (this.U[k].length === j) this.U[k].push(0);
        this.U[k][j] += 1;
      }
    }

    for (let j = 0; j < row.reduce((a, b) => a + b, 0); j++) {
      if (this.V.length === j) this.V.push(0);
      this.V[j] += 1;
    }
  }
}

function dirichLogProb(priorList: number[], data: CompressedRowData): number {
  let K = data.K;
  let total = 0.0;
  
  for (let k = 0; k < K; k++) {
    for (let i = 0; i < data.U[k].length; i++) {
      total += data.U[k][i]*Math.log(priorList[k] + i);
    }
  }

  let sumPrior = priorList.reduce((a, b) => a + b, 0);
  for (let i = 0; i < data.V.length; i++) {
    total -= data.V[i] * Math.log(sumPrior + i);
  }

  return total;
}

function priorGradient(priorList: number[], data: CompressedRowData): number[] {
  let K = data.K;
  let sumPrior = sum(priorList);

  let termToSubtract = 0;
  for (let i = 0; i < data.V.length; i++) {
    termToSubtract += data.V[i] / (sumPrior + i);
  }

  let retVal: number[] = new Array(K).fill(0);
  for (let j = 0; j < K; j++) {
    for (let i = 0; i < data.U[j].length; i++) {
      retVal[j] += data.U[j][i] / (priorList[j] + i);
    }
  }

  for (let k = 0; k < K; k++) {
    retVal[k] -= termToSubtract;
  }

  return retVal;
}

function sum(arr: number[]): number {
  return arr.reduce((a, b) => a + b, 0);
}

function priorHessianConst(priorList: number[], data: CompressedRowData): number {
  return sum(data.V.map((v, i) => v / Math.pow(sum(priorList) + i, 2)));
}

function priorHessianDiag(priorList: number[], data: CompressedRowData): number[] {
  let K = data.U.length;
  let retVal: number[] = new Array(K).fill(0);
  
  for (let k = 0; k < K; k++) {
    for (let i = 0; i < data.U[k].length; i++) {
      retVal[k] -= data.U[k][i] / Math.pow(priorList[k] + i, 2);
    }
  }

  return retVal;
}

// Compute the next value to try here
// http://research.microsoft.com/en-us/um/people/minka/papers/dirichlet/minka-dirichlet.pdf (eq 18)
//https://www.researchgate.net/profile/Max-Sklar/publication/370927162_Algorithms_for_Multivariate_Newton-Raphson_for_Optimization/links/64698c0170202663165fd82a/Algorithms-for-Multivariate-Newton-Raphson-for-Optimization.pdf
function getPredictedStep(hConst: number, hDiag: number[], gradient: number[]): number[] {
  let K = gradient.length;
  let numSum = sum(gradient.map((g, k) => g / hDiag[k]));
  let denSum = sum(hDiag.map(h => 1.0 / h));
  let b = numSum / ((1.0 / hConst) + denSum);
  return gradient.map((g, k) => (b - g) / hDiag[k]);
}

// Uses the diagonal hessian on the log-alpha values
// https://www.researchgate.net/profile/Max-Sklar/publication/370927162_Algorithms_for_Multivariate_Newton-Raphson_for_Optimization/links/64698c0170202663165fd82a/Algorithms-for-Multivariate-Newton-Raphson-for-Optimization.pdf
function getPredictedStepAlt(hConst: number, hDiag: number[], gradient: number[], alphas: number[]): number[] {
  let x = gradient.map((grad, i) => grad + alphas[i] * hDiag[i]);
  let Z = 1.0 / hConst + sum(alphas.map((alpha, i) => alpha / x[i]));
  let S = sum(alphas.map((alpha, i) => alpha * gradient[i] / x[i]));
  return gradient.map((grad, i) => (S / Z - grad) / x[i]);
}

//The priors and data are global, so we don't need to pass them in
function getTotalLoss(trialPriors: number[], data: CompressedRowData): number {
  return -1 * dirichLogProb(trialPriors, data);
}

function predictStepUsingHessian(gradient: number[], priors: number[], data: CompressedRowData): number[] {
  let totalHConst = priorHessianConst(priors, data);
  let totalHDiag = priorHessianDiag(priors, data);
  return getPredictedStep(totalHConst, totalHDiag, gradient);
}

function predictStepLogSpace(gradient: number[], priors: number[], data: CompressedRowData): number[] {
  let totalHConst = priorHessianConst(priors, data);
  let totalHDiag = priorHessianDiag(priors, data);
  return getPredictedStepAlt(totalHConst, totalHDiag, gradient, priors);
}

// Returns whether it's a good step, and the loss
function testTrialPriors(trialPriors: number[], data: CompressedRowData): number {
  if (trialPriors.some(alpha => alpha <= 0)) {
  return Infinity;
  }

  return getTotalLoss(trialPriors, data);
}

function sqVectorSize(v: number[]): number {
  return sum(v.map(x => x ** 2));
}

export function findDirichletPriors(data: CompressedRowData, iterations: number): number[] {
  let priors = Array.from({length: data.K}, () => 1.0 / data.K);

  // Let the learning begin!!
  // Only step in a positive direction, get the current best loss.
  let currentLoss = getTotalLoss(priors, data);

  let gradientToleranceSq = 2 ** -10;
  let learnRateTolerance = 2 ** -20;

  let count = 0;
  while(count < iterations) {
    count += 1;

    // Get the data for taking steps
    let gradient = priorGradient(priors, data);
    let gradientSize = sqVectorSize(gradient); 
    console.debug(`Iteration: ${count} Loss: ${currentLoss} ,Priors: ${priors}, Gradient Size: ${gradientSize}`);

    if (gradientSize < gradientToleranceSq) {
      console.debug("Converged with small gradient");
      return priors;
    }

    let trialStep = predictStepUsingHessian(gradient, priors, data);

    // First, try the second order method
    let trialPriors = priors.map((prior, i) => prior + trialStep[i]);

    // TODO: Check for taking such a small step that the loss change doesn't register (essentially converged)
    //  Fix by ending

    let loss = testTrialPriors(trialPriors, data);
    if (loss < currentLoss) {
      currentLoss = loss;
      priors = trialPriors;
      continue;
    }

    trialStep = predictStepLogSpace(gradient, priors, data);
    trialPriors = priors.map((prior, i) => {
      try {
        return prior * Math.exp(trialStep[i]);
      } catch (e) {
        return prior;
      }
    });

    loss = testTrialPriors(trialPriors, data);

    // Step in the direction of the gradient until there is a loss improvement
    let learnRate = 1.0;
    while (loss > currentLoss) {
      learnRate *= 0.9;
      trialPriors = priors.map((prior, i) => prior + gradient[i]*learnRate);
      loss = testTrialPriors(trialPriors, data);
    }

    if (learnRate < learnRateTolerance) {
      console.debug("Converged with small learn rate");
      return priors;
    }

    currentLoss = loss;
    priors = trialPriors;
  }

  console.debug("Reached max iterations");
  return priors;
}